lib/fastly_nsq/manager.rb

```diff
@@ -1,104 +1,80 @@
-  attr_reader :listeners
+  DEADLINE = 30
+  DEFAULT_POOL_SIZE = 5
+
+  attr_reader :done, :pool, :logger
 
-  def initialize(options = {})
-    @options = options
-    @done = false
-    @listeners = Set.new
-    @plock = Mutex.new
+  def initialize(logger: FastlyNsq.logger, **pool_options)
+    @done      = false
+    @logger    = logger
+    @pool      = FastlyNsq::PriorityThreadPool.new(
+      { fallback_policy: :caller_runs, max_threads: DEFAULT_POOL_SIZE }.merge(pool_options),
+    )
   end
 
-  def start
-    setup_configured_listeners
-    @listeners.each(&:start)
+  def topic_listeners
+    @topic_listeners ||= {}
   end
 
-  def quiet
-    return if @done
-    @done = true
+  def topics
+    topic_listeners.keys
+  end
 
-    FastlyNsq.logger.info { 'Terminating quiet listeners' }
-    @listeners.each(&:terminate)
+  def listeners
+    topic_listeners.values
   end
 
-  PAUSE_TIME = 0.5
+  def terminate(deadline = DEADLINE)
+    return if done
 
-  def stop(deadline)
-    quiet
+    stop_listeners
 
-    sleep PAUSE_TIME
-    return if @listeners.empty?
+    return if pool.shutdown?
 
-    FastlyNsq.logger.info { 'Pausing to allow workers to finish...' }
-    remaining = deadline - Time.now
-    while remaining > PAUSE_TIME
-      return if @listeners.empty?
-      sleep PAUSE_TIME
-      remaining = deadline - Time.now
-    end
-    return if @listeners.empty?
+    stop_processing(deadline)
 
-    hard_shutdown
+    @done = true
   end
 
   def stopped?
-    @done
+    done
   end
 
-  def listener_stopped(listener)
-    @plock.synchronize do
-      @listeners.delete listener
+  def add_listener(listener)
+    logger.info { "Listening to topic:'#{listener.topic}' on channel: '#{listener.channel}'" }
+
+    if topic_listeners[listener.topic]
+      logger.warn { "topic: #{listener.topic} was added more than once" }
     end
+
+    topic_listeners[listener.topic] = listener
   end
 
-  def listener_killed(listener)
-    @plock.synchronize do
-      @listeners.delete listener
-      unless @done
-        FastlyNsq.logger.info { "recreating listener for: #{listener.identity}" }
-        new_listener = listener.reset_then_dup
-        @listeners << new_listener
-        new_listener.start
-      end
-    end
+  def transfer(new_manager, deadline: DEADLINE)
+    new_manager.topic_listeners.merge!(topic_listeners)
+    stop_processing(deadline)
+    topic_listeners.clear
+    @done = true
   end
 
-  private
+  protected
 
-  def setup_configured_listeners
-    FastlyNsq.logger.debug { "options #{@options.inspect}" }
-    FastlyNsq.logger.debug { "starting listeners: #{FastlyNsq.topic_map.inspect}" }
+  def stop_processing(deadline)
+    logger.info { 'Stopping processors' }
+    pool.shutdown
 
-    FastlyNsq.topic_map.each_pair do |topic, processor|
-      @listeners << setup_listener(topic, processor)
-    end
-  end
+    logger.info { 'Waiting for processors to finish...' }
+    return if pool.wait_for_termination(deadline)
```

Comment:
does `wait_for_termination` return a boolean?  If it returns true, what does that mean for the `transfer` method above?  it seems like transfer will continue assuming `stop_processing` is done.

Reply:
Yeah [wait_for_termination](http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/ThreadPoolExecutor.html#wait_for_termination-instance_method) returns `true` if the threads were terminated within the timeout, `false` otherwise.

Reply:
ah, makes sense now. terminate gracefully or force terminate.